recipes/Libriheavy/libriheavy_prepare.py

```diff
@@ -0,0 +1,402 @@
+    different native speakers of English.
+    Sometimes this is because the word with the accent is foreign, but often the
+    other words in the Sentence are also hard to pronounce.
+    An extreme example is ö,ß, and ü, which are often whole German sentences.
+    CommonVoice regularly has data added to it.
+    This function may error out if new characters show up in the training, dev,
+    or test sets.
+    If this happens, add the case to test_common_voice_prepare.py, and fix this
+    function in a backward-compatible manner.
+    """
+
+    # These characters mean we should discard the sentence, because the
+    # pronunciation will be too uncertain.
+    stop_characters = (
+        "["
+        "áÁàăâåäÄãÃāảạæćčČçÇðéÉèÈêěëęēəğíîÎïīịıłṃńňñóÓòôőõøØōŌœŒřšŠşșȘúÚûūụýžþ"
+        # Suggests the sentence is not English but German.
+        "öÖßüÜ"
+        # All sorts of languages: Greek, Arabic...
+        "\u0370-\u1AAF"
+        # Chinese/Japanese/Korean.
+        "\u4E00-\u9FFF"
+        # Technical symbols.
+        "\u2190-\u23FF"
+        # Symbols that could be pronounced in various ways.
+        "\\[\\]€→=~%§_#"
+        "]"
+    )
+
+    # These characters mark word boundaries.
+    split_character_regex = '[ ",:;!?¡\\.…()\\-—–‑_“”„/«»]'
+
+    # These could all be used as apostrophes in the middle of words.
+    # If at the start or end of a word, they will be removed.
+    apostrophes_or_quotes = "['`´ʻ‘’]"
+
+    sentence_level_mapping = {"&": " and ", "+": " plus ", "ﬂ": "fl"}
+
+    final_characters = set(" ABCDEFGHIJKLMNOPQRSTUVWXYZ'")
+
+    if not re.search(stop_characters, sentence) is None:
+        return None
+
+    sentence_mapped = sentence
+    if any((source in sentence) for source in sentence_level_mapping):
+        for source, target in sentence_level_mapping.items():
+            sentence_mapped = sentence_mapped.replace(source, target)
+
+    # Some punctuation that indicates a word boundary.
+    words_split = re.split(split_character_regex, sentence_mapped)
+    words_quotes = [
+        # Use ' as apostrophe.
+        # Remove apostrophes at the start and end of words (probably quotes).
+        # Word-internal apostrophes, even where rotated, are retained.
+        re.sub(apostrophes_or_quotes, "'", word).strip("'")
+        for word in words_split
+    ]
+
+    # Processing that does not change the length.
+    words_upper = [word.upper() for word in words_quotes]
+
+    words_mapped = [
+        # word.translate(character_mapping)
+        word
+        for word in words_upper
+        # Previous processing may have reduced words to nothing.
+        # Remove them.
+        if word != ""
+    ]
+
+    result = " ".join(words_mapped)
+    character_set = set(result)
+    assert character_set <= final_characters, (
+        "Unprocessed characters",
+        sentence,
+        result,
+        character_set - final_characters,
+    )
+    return result
```

Comment:
Random note @TParcollet, at some point we should consider having our own TextNormalizer. It could be helpful in the case of preprocessing dataset / normalising output of ASR systems for instance.